include/FNA3D.h

```diff
@@ -1538,6 +1549,112 @@ FNA3DAPI void FNA3D_SetStringMarker(FNA3D_Device *device, const char *text);
+
+typedef struct SDL_GpuGraphicsShaderInfo SDL_GpuGraphicsShaderInfo;
+typedef struct SDL_GpuComputeShaderInfo SDL_GpuComputeShaderInfo;
+typedef struct SDL_GpuComputeBufferBinding SDL_GpuComputeBufferBinding;
+typedef struct SDL_GpuComputeTextureBinding SDL_GpuComputeTextureBinding;
+
+/* Binds a vertex-fragment shader pair to be used with draw calls.
+ * If you call ApplyEffect, the shaders will be treated as unbound
+ * and it will become invalid to call any of the graphics shader EXT functions.
+ *
+ * vertShaderInfo: A struct containing the vertex shader module to bind and relevant metadata.
+ * fragShaderInfo: A struct containing the fragment shader module to bind and relevant metadata.
+ */
+FNA3DAPI void FNA3D_BindGraphicsShadersEXT(
+	FNA3D_Device *device,
+	SDL_GpuGraphicsShaderInfo *vertShaderInfo,
+	SDL_GpuGraphicsShaderInfo *fragShaderInfo
+);
+
+/* Pushes vertex uniform data to be used with subsequent draw calls.
+ * You may only call this after FNA3D_BindGraphicsShadersEXT has been called.
+ *
+ * data: client data to write into the uniform buffer
+ * dataLengthInBytes: the length of the client data to write
+ */
+FNA3DAPI void FNA3D_PushVertexShaderUniformsEXT(
+	FNA3D_Device *device,
+	void *data,
+	uint32_t dataLengthInBytes
+);
+
+/* Pushes fragment uniform data to be used with subsequent draw calls.
+ * You may only call this after FNA3D_BindGraphicsShadersEXT has been called.
+ *
+ * data: client data to write into the uniform buffer
+ * dataLengthInBytes: the length of the client data to write
+ */
+FNA3DAPI void FNA3D_PushFragmentShaderUniformsEXT(
+	FNA3D_Device *device,
+	void *data,
+	uint32_t dataLengthInBytes
+);
+
+/* Binds a compute shader to be used with compute dispatch calls.
+ * Note that this will break the render pass and unbind the graphics shaders,
+ * so interleaving this with draw calls will have bad performance.
+ *
+ * computeShaderInfo: A struct containing the compute shader module to bind and relevant metadata.
+ */
+FNA3DAPI void FNA3D_BindComputeShaderEXT(
+	FNA3D_Device *device,
+	SDL_GpuComputeShaderInfo *computeShaderInfo
+);
+
+/* Binds buffers for use with the bound compute pipeline.
+ * You may only call this function after FNA3D_BindComputeShaderEXT has been called.
+ *
+ * pBindings: An array of structs containing buffer and cycle option pairs.
+ *            Must be at least as long as the buffer bindings specified in the compute shader.
+ */
+FNA3DAPI void FNA3D_BindComputeBuffersEXT(
+	FNA3D_Device *device,
+	SDL_GpuComputeBufferBinding *pBindings
+);
+
+/* Binds textures for use with the bound compute pipeline.
+ * You may only call this function after FNA3D_BindComputeShaderEXT has been called.
+ *
+ * pBindings: An array of structs containing texture and cycle option pairs.
+ *            Must be at least as long as the texture bindings specified in the compute shader.
+ */
+FNA3DAPI void FNA3D_BindComputeTexturesEXT(
+	FNA3D_Device *device,
+	SDL_GpuComputeTextureBinding *pBindings
```

Comment:
Is there a blessed way to feed FNA3D textures in to this? Or would you use a SysTexture?
From looking at this it seems like in this proposal, fragment/vertex shaders ('graphics' shaders) can only consume FNA3D textures, and compute shaders can only consume raw SDL_Gpu textures.

Reply:
FNA3D texture handles _are_ just sdl_gpu texture handles, but I agree that this is awkward... I'll just wrap the SDL struct and handle it in the implementation

Reply:
Actually I was wrong and FNA3D_Driver_SDL was wrapping the texture handles. So nice catch!